File: Sliding Window Maximum/sliding window maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> answers;
        deque<int> de;
        de.push_back(0);
        for(int i = 1; i < k; i++) {
            while (de.size() > 0 && nums[de.back()] < nums[i]) {
                de.pop_back();
            }
            de.push_back(i);
        }
        answers.push_back(nums[de.front()]);
        
        for(int j = k; j < nums.size(); j++) {
            int startingPoint = j - k + 1;
            while (de.size() > 0 && de.front() < startingPoint) {
                de.pop_front();
            }
            while (de.size() > 0 && nums[de.back()] < nums[j]) {
                de.pop_back();
            }
            de.push_back(j);
            answers.push_back(nums[de.front()]);
        }
        
        return answers;
    }
};

File: Sliding Window Maximum/sliding window maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        answers = []
        n = len(nums)
        de = Deque()
        de.append(0)
        for i in range(1, k):
            while (len(de) > 0 and nums[de[-1]] < nums[i]):
                de.pop()
            de.append(i)
        answers.append(nums[de[0]])
        
        for j in range(k, n):
            startingPoint = j - k + 1
            while (len(de) > 0 and de[0] < startingPoint):
                de.popleft()
            
            while (len(de) > 0 and nums[de[-1]] < nums[j]):
                de.pop()
            
            de.append(j)
            answers.append(nums[de[0]])
        
        return answers